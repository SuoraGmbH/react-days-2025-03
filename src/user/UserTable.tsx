import { FunctionComponent, useEffect, useState } from "react";
import { Loader, XCircle, ChevronUp, ChevronDown } from "lucide-react";

interface User {
  id: number;
  name: string;
  email: string;
  company: {
    name: string;
  };
  address: {
    city: string;
  };
}

export const UserTable: FunctionComponent = () => {
  const [users, setUsers] = useState<User[]>([]);
  const [loading, setLoading] = useState(false);
  const [error, setError] = useState<string | null>(null);
  const [filterLetter, setFilterLetter] = useState("All");
  const [sortAsc, setSortAsc] = useState(true);

  useEffect(() => {
    const fetchUsers = async () => {
      setLoading(true);
      setError(null);

      try {
        const response = await fetch(
          "https://jsonplaceholder.typicode.com/users",
        );
        if (!response.ok) {
          throw new Error("Failed to fetch users.");
        }

        const data = await response.json();
        setUsers(data);
      } catch (e) {
        setError(e instanceof Error ? e.message : "Unknown error");
      } finally {
        setLoading(false);
      }
    };

    fetchUsers();
  }, []);

  const filteredUsers = users.filter(
    (user) => filterLetter === "All" || user.name.startsWith(filterLetter),
  );

  const sortedUsers = [...filteredUsers].sort((a, b) =>
    sortAsc ? a.name.localeCompare(b.name) : b.name.localeCompare(a.name),
  );

  return (
    <div className="p-6 space-y-4">
      <h1 className="text-2xl font-bold">User Dashboard</h1>

      {/* Loading and Error State */}
      {loading ? (
        <div className="flex justify-center items-center text-blue-500">
          <Loader className="animate-spin" size={24} />
          <span className="ml-2">Loading users...</span>
        </div>
      ) : error ? (
        <div className="flex items-center text-red-500">
          <XCircle size={24} />
          <span className="ml-2">{error}</span>
        </div>
      ) : (
        <>
          {/* Filters */}
          <div className="flex space-x-2">
            {["All", "A", "B", "C", "D"].map((letter) => (
              <button
                key={letter}
                onClick={() => setFilterLetter(letter)}
                className={`px-3 py-2 rounded border ${
                  filterLetter === letter
                    ? "bg-blue-500 text-white"
                    : "border-gray-300 bg-white text-black hover:bg-gray-100"
                }`}
              >
                {letter}
              </button>
            ))}
          </div>

          {/* Sorting */}
          <button
            onClick={() => setSortAsc(!sortAsc)}
            className="mt-2 px-3 py-2 border border-gray-300 rounded bg-gray-100 hover:bg-gray-200"
          >
            Sort by Name{" "}
            {sortAsc ? <ChevronUp size={16} /> : <ChevronDown size={16} />}
          </button>

          {/* User Table */}
          {sortedUsers.length > 0 ? (
            <div className="mt-4 border border-gray-300 rounded overflow-hidden">
              <table className="w-full border-collapse">
                <thead>
                  <tr className="bg-gray-100 text-left">
                    <th className="px-4 py-2 border-b">ID</th>
                    <th className="px-4 py-2 border-b">Name</th>
                    <th className="px-4 py-2 border-b">Email</th>
                    <th className="px-4 py-2 border-b">Company</th>
                    <th className="px-4 py-2 border-b">City</th>
                  </tr>
                </thead>
                <tbody>
                  {sortedUsers.map((user) => (
                    <tr key={user.id} className="border-b">
                      <td className="px-4 py-2">{user.id}</td>
                      <td className="px-4 py-2">{user.name}</td>
                      <td className="px-4 py-2">{user.email}</td>
                      <td className="px-4 py-2">{user.company.name}</td>
                      <td className="px-4 py-2">{user.address.city}</td>
                    </tr>
                  ))}
                </tbody>
              </table>
            </div>
          ) : (
            <p className="text-gray-500 mt-4">No users found.</p>
          )}
        </>
      )}
    </div>
  );
};
