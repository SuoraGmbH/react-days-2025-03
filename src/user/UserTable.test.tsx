import { describe, expect, test } from "vitest";
import { UserTable } from "./UserTable.tsx";
import { render, screen } from "@testing-library/react";

describe("<UserTable>", () => {
  test("renders without crashing", async () => {
    render(<UserTable />);
  });

  test("loading spinner matches snapshot", async () => {
    const result = render(<UserTable />);

    expect(result.container).toMatchSnapshot();
  });

  test("rendered table matches snapshot", async () => {
    const result = render(<UserTable />);

    await screen.findByText("Chelsey Dietrich");

    expect(result.container).toMatchSnapshot();
  });
});
