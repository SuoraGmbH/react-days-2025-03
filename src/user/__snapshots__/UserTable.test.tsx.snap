// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<UserTable> > loading spinner matches snapshotâ€° 1`] = `
<div>
  <div
    class="p-6 space-y-4"
  >
    <h1
      class="text-2xl font-bold"
    >
      User Dashboard
    </h1>
    <div
      class="flex justify-center items-center text-blue-500"
    >
      <svg
        class="lucide lucide-loader animate-spin"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M12 2v4"
        />
        <path
          d="m16.2 7.8 2.9-2.9"
        />
        <path
          d="M18 12h4"
        />
        <path
          d="m16.2 16.2 2.9 2.9"
        />
        <path
          d="M12 18v4"
        />
        <path
          d="m4.9 19.1 2.9-2.9"
        />
        <path
          d="M2 12h4"
        />
        <path
          d="m4.9 4.9 2.9 2.9"
        />
      </svg>
      <span
        class="ml-2"
      >
        Loading users...
      </span>
    </div>
  </div>
</div>
`;
